### ISSUE TITLE
Hillshade: severe performance degradation when crossing antimeridian

### ISSUE THREAD

**Issue by jessevarner** [Esri/raster-functions#35]
I've been trying out the Hillshade function on mosaic datasets containing bathymetric data. I really like how the Python version looks vs. the built-in mosaic function -- it's much smoother-looking. However, I'm running into a problem when the viewing window crosses the antimeridian. The drawing performance gets severely degraded, and it may take several minutes for the image to display. This becomes an issue when the mosaic is published as an image service to be used in a map viewer, and the user zooms into an area surrounding 180 longitude in the Pacific.

I tried tracing what was going on by using print statements in the Hillshade.py code. I can see that, instead of the updatePixels() function getting called once, it gets called dozens or hundreds of times in a row, with what seems to be tiny windows of pixels.

Here are some steps to replicate the problem:
1. Add 2 rasters to a mosaic dataset, on either side of the antimeridian. Here are some example files (MRF format, LERC compression): https://drive.google.com/file/d/0Bz0z_q8d9UpDYzhBeUFRd1dKdG8/view?usp=sharing
2. Add the Python Hillshade function into the raster function chain.
3. Change the map's projection to something Pacific-centered, such as PDC Mercator.
4. Zoom in near the antimeridian, and it will take a very long time to display.

Not sure if this is a problem with the mosaic dataset code in ArcGIS, or with the Python function. I'm using ArcMap 10.4.1.

Thanks,
Jesse

**Comment by jdrisdelle** [comment]
Hi Jesse,

I was able to reproduce your issues. I don't think it is specific to the hillshade function, It looks like it is inherent to the Python raster function API. I am going to have one of my developers take a look at it.  I'll let you know when i have more information. Sorry for the inconvenience.

**Comment by jdrisdelle** [comment]
Hi Jesse, 

It seems like setting the ‘resampling’ flag to False in .getConfiguration() method is causing the performance issue. Could you try setting this flag to True instead of False.

Setting the resampling flag to False passes the pixel values at native resolution to the python script in .updatePixels() and ArcGIS then resamples this pixel block which in turn invokes .updatePixels() several times.Setting it to True passes the already resampled pixel values to python script. Thus .updatePixels() is invoked only once.

Let me know if this solves the issue.  If this works for you i can update the Python code in the GitHub Repo.

**Comment by jessevarner** [comment]
Hi Jamie,
Some observations after setting the resampling flag to True:

1. When I zoom way into the image, the image looks more pixelated and "chunkier", whereas if you set resampling:True, the image looks smoother as you zoom in past the raster resolution. That's the main reason I was hoping to use the Python Hillshade instead of the built-in Hillshade raster function. Here are a couple screenshots:

resampling:True:
![image](https://cloud.githubusercontent.com/assets/22140704/22571888/fb2aea50-e95e-11e6-8e0e-cf531eb8bc22.png)

resampling:False:
![image](https://cloud.githubusercontent.com/assets/22140704/22571896/0624a9c8-e95f-11e6-8341-15283885060f.png)

2. The Z factor gives unexpected results if I switch to a projected coordinate system. If I set the Z factor to 1, it looks OK in WGS84, but if I switch to Web Mercator, it gets overly exaggerated. The same doesn't happen with resampling:False. Screenshot of what I'm seeing: 
![image](https://cloud.githubusercontent.com/assets/22140704/22572138/f41e548a-e95f-11e6-9de9-dfc434129d99.png)

3. If my viewing window crosses the antimeridian (in PDC Mercator for example), I'm just getting a black image. When I scroll so the antimeridian isn't in view, it does display the hillshade correctly (ignoring the Z factor problem in #2).

4. Also, if I'm in Web Mercator, I'm getting an error when I zoom out past a certain level. Here's the error:
![image](https://cloud.githubusercontent.com/assets/22140704/22572192/27beba1e-e960-11e6-8e3c-4bd3d8afa8d1.png)

Thanks for looking into this!

Cheers,
Jesse

**Comment by jdrisdelle** [comment]
Hi Jess, 

As far as the resampling goes...Try setting resampling at the mosaic dataset level.  Right click on it in Arc Catalog and open it's properties.  Play around with the default resampling property in the defaults tab.  I will get back to you on the other issues once i figure them out.

**Comment by jdrisdelle** [comment]
Hi Jesse, 

Just wanted to let you know that we are still investigating this.  I'll let you know when we figure this out.  Sorry for the delay.

**Comment by jdrisdelle** [comment]
@jessevarner , 

Can you test the fixes in the following branch.  I tried it with your data and everything looks good now. I just want to make sure that everything is ok on your end before i commit it to the master branch. You jsut need to grab the utils.py and hillshade.py

https://github.com/Esri/raster-functions/tree/Hillshade-fixes

**Comment by jessevarner** [comment]
Hi Jamie,
Thanks, this does solve the performance issue when displaying across the antimeridian, as well as the excessive vertical exaggeration I was seeing. 

I am still seeing "chunky" pixels when I zoom way in, no matter what resampling method I choose for the mosaic dataset. But that's not a major issue... the biggest problem was the performance. Thank you for resolving this issue!

Cheers,
Jesse

**Comment by jdrisdelle** [comment]
No problem. Are you just zooming into the pixel resolution?  That may be a by-product of the data itself.

**Comment by jessevarner** [comment]
Yes, I'm talking about zooming in past the pixel resolution. If I set resampling=False in the script, it looks smooth as you keep zooming in, but with resampling=True, it ends up looking "chunky" no matter what resampling I use on the mosaic dataset or in the display properties. Here's an example for comparison:

resampling=True
![image](https://cloud.githubusercontent.com/assets/22140704/23484738/686c40be-fe95-11e6-9d59-1b38feca2e69.png)

resampling=False
![image](https://cloud.githubusercontent.com/assets/22140704/23484743/709730c8-fe95-11e6-834f-411a6f00937c.png)

I prefer the look of the second one. Not really a major issue, but just something I noticed.
